### PR TITLE
Flesh out synchronization of session span transition

### PR DESCRIPTION
**Pull request by bidetofevil**
## Goal

Reviewing synchronization of the current span object and tweaked  it to prevent potentially double init. While it really wouldn't happen now, who knows what the future brings.

### CODE REVIEW

**Comment by bidetofevil** [comment]
* **#937** <a href="https://app.graphite.dev/github/pr/embrace-io/embrace-android-sdk/937?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @bidetofevil and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/spans/CurrentSessionSpanImpl.kt [hunk 1]
```diff
@@ -36,8 +36,12 @@ internal class CurrentSessionSpanImpl(
     private val sessionSpan: AtomicReference<PersistableEmbraceSpan?> = AtomicReference(null)
 
     override fun initializeService(sdkInitStartTimeMs: Long) {
-        synchronized(sessionSpan) {
-            sessionSpan.set(startSessionSpan(sdkInitStartTimeMs))
+        if (sessionSpan.get() == null) {
+            synchronized(sessionSpan) {
+                if (sessionSpan.get() == null) {
```
Comment (fractalwrench) [inline embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/spans/CurrentSessionSpanImpl.kt:41]: Is it worth formalising this pattern in a separate PR? It's raised a couple of questions in the past - I understand the rationale, but others might not

```
private inline fun doubleCheckLock(lock: Any, predicate: () -> Boolean, action: () -> Unit) {
        if (predicate()) {
            return
        }
        synchronized(lock) {
            if (predicate()) {
                return
            }
            action()
        }
    }
```

**Review by fractalwrench — APPROVED** [review]
LGTM

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/spans/CurrentSessionSpanImpl.kt (hunk 1 above)
Comment (bidetofevil) [reply embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/spans/CurrentSessionSpanImpl.kt:41]: Yeah we should! We'll have to be careful how we implement these - we don't want a dangling reference in an `AtomicReference` causing leaks.

**Pull request merged**